### ISSUE TITLE
Freeze while clicking/moving the title bar

### ISSUE THREAD

**Issue by Scellow** [MonoGame/MonoGame#5656]
Hello

Currently each time you click or hold the mouse button or drag the title bar the window will freeze

I don't think that should happen, any idea ?

**Comment by Jjagg** [comment]
That's expected behaviour. The same happens in XNA.

**Comment by Scellow** [comment]
It should be treated as a bug imo, this is really annoying specially in networked games

**Comment by Jjagg** [comment]
I agree it would be nice to not block the game. I'll play around a bit to try and fix it. Currently doing some stuff for DX, so that'll be first. @tomspilman Do you agree we should treat this as a bug?

**Comment by tomspilman** [comment]
This sounds similar to the window resizing discussion we had before.  In XNA the window updates were disabled during resize or window dragging.  I think we agreed if we wanted to have the window continue to update during resize (or dragging in this case) that we want to make it something the developer opts-in for.

**Comment by Jjagg** [comment]
This is different because the game logic is completely blocked in this case, so update isn't even called. The issue is that [`Application.DoEvents`](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Windows/WinFormsGameWindow.cs#L340) blocks while the user is resizing or dragging the window. We could use a timer to make sure `Game.Tick` is still called during these operations. I would prefer this to happen by default, but if it's opt-in that's fine too.
It would be good to at least have some way to keep the game going, because like @Scellow said this can be an issue for some types of games.

**Comment by tomspilman** [comment]
> This is different because the game logic is completely blocked in this case

You sure Update isn't also blocked in XNA on window dragging?

**Comment by Jjagg** [comment]
I phrased that wrong. Yeah it is, but this is a different thing. This is about the run loop blocking when dragging/resizing while the other discussion was about resizing the backbuffer during window resizing. There's an argument for not resizing the backbuffer every frame, but is there for blocking the run loop during resize/draw? There might be conflicts if you try to change window size in your game logic during the dragging operation, but that can be worked around. Other than that I don't see any downsides.

**Comment by cra0zy** [comment]
On Mac/Linux the window contents does not get frozen while you drag the window.

**Comment by Jjagg** [comment]
That might make it hard to get right on Windows under SDL :/
Or maybe not, since it's Win32 in the backend we could maybe just do something like
```
if (event is resize start && platform is windows)
{
    Start Win32 timer like in DX projects
}
if (event is resize end && platform is windows)
{
    Kill Win32 timer
}
```
in the event handling code.

**Comment by Apostolique** [comment]
Any news on this? It's really annoying that the game completely pauses when [clicking / holding click on title bar / dragging the window around].

**Comment by tomspilman** [comment]
> Any news on this?

I am certain it can be done.  But no one has spent the time yet implementing the feature under the Windows DirectX platform.

**Comment by squarebananas** [comment]
Noticed this post just and was planning on having a look at this anyway this week. It seems that anything under Game.Run including the run loop is blocked while title bar is held. As far as I could tell this block seems to happen due to the form's behaviour rather than anything Monogame is overriding (unless I missed something).

I tried this just as a temporary solution and it seemed to work. However I'm not 100% confident this will cover everything and really it is just proof of concept showing it is at least possible. It can probably be achieved in a much better way.

To try it out make these changes in WinFormsGameWindow.cs


Add the following next to the existing event handlers inside the WinFormsGameWindow constructor
`Form.ResizeBegin += OnResizeBegin;`

Add the following at the start of the OnResizeEnd method
`windowMovingUpdatesRequired = false;`

Change the following inside the RunLoop method
`UpdateWindows();`
`Game.Tick `
to this
`TryGameTick();`

Finally place this inside the class
```bool windowMovingUpdatesRequired = false;
public async void OnResizeBegin(object sender, EventArgs eventArgs)
{
    windowMovingUpdatesRequired = true;
    while (windowMovingUpdatesRequired == true)
    {
        TryGameTick();
        await System.Threading.Tasks.Task.Delay(1);
    }
}

bool gameTickAlreadyRunning = false;
public void TryGameTick()
{
    if (gameTickAlreadyRunning == false)
    {
        gameTickAlreadyRunning = true;
        UpdateWindows();
        Game.Tick();
        gameTickAlreadyRunning = false;
    }
 }
```

Basically this works by attempting to run Game.Tick inside a new loop after OnResizeBegin is called (which occurs when the title bar is moved). This will continue until OnResizeEnd is called and then the Game.Tick calls will continue from RunLoop instead. I'm not sure if the TryGameTick method is actually necessary but I've left it in as a precaution against starting Game.Tick again before it has ended.

One thing I have noticed is grabbing the title bar without moving the mouse at all will cause a freeze for half a second anyway. If the mouse is moved just slightly though this freeze will end early.

**Comment by Apostolique** [comment]
> I am certain it can be done. But no one has spent the time yet implementing the feature under the Windows DirectX platform.

It also happens in Windows OpenGL.

btw good work @squarebananas

**Comment by nkast** [comment]
The stackexchange link posted by @Apostolique  has some more links about this, this one explains a lot about the issue: https://www.gamedev.net/forums/topic/488074-win32-message-pump-and-opengl---rendering-pauses-while-draggingresizing/?tab=comments#comment-4189300

There are only two thinks you can do. Run the gameloop on it's own thread, or handle move/resize messages internally, (don't pass them down to DefWindowProc() ),

**Comment by nkast** [comment]
One other thing I noticed, is that if you resize the window you will see a while background filling the rest of the window. This looks back and just makes people wonder if something is wrong. 
XNA would stretch content of the window, no idea how it does that. 

I tried a couple of things, setting  the swapchain to Stretched or Width/Height to 0.
The content wont get refreshed until the next present().

**Comment by nkast** [comment]
here:
https://github.com/MonoGame/MonoGame/blob/2290384ccf9dfe9e45bca166cad7cc146f4c2bc7/MonoGame.Framework/Windows/WinFormsGameWindow.cs#L348-L351

I replaced it with: 
```
            if (_switchingFullScreen || Form.IsResizing)
            {
                // repaint the window
                this.Game.GraphicsDevice.Present();
                return;
            }
```

@tomspilman , @Jjagg 
What can go wrong with that?

**Comment by cra0zy** [comment]
I think I have a good idea on how to fix this for the SDL side.

**Comment by cra0zy** [comment]
I thought SDL_HINT_WINDOWS_ENABLE_MESSAGELOOP would be helpful, but nope :(

**Comment by MichaelDePiazzi** [comment]
I would also love a solution to this. For networked multiplayer games like mine, this can cause some issues. I'm liking @squarebananas idea of temporarily running the game loop on another thread while the resize is occurring. It definitely seems promising.

**Comment by Apostolique** [comment]
@MichaelDePiazzi The solution I have in my game for multiplayer is to not use the default monogame GameTime. I use StopWatch from System.Diagnostics instead. This one doesn't get out of sync no matter what.

**Comment by Apostolique** [comment]
Related issue: #5590

**Comment by MichaelDePiazzi** [comment]
FYI - I've tried out @squarebananas solution, and it seems to work quite well! I'll be releasing an experimental build of my game soon to test it out further.

I did tweak it slightly and used a timer to avoid the allocations from `Task.Delay` though. I also didn't bother with `TryGameTick`. It didn't seem necessary with this timer approach given everything runs on the same thread.

My version looks like this:
```csharp
private void OnResizeBegin(object sender, EventArgs e)
{
    _isResizeTickEnabled = true;
    _resizeTickTimer.Enabled = true;
}

private void OnResizeTick(object sender, System.Timers.ElapsedEventArgs e)
{
    if (!_isResizeTickEnabled)
        return;
    UpdateWindows();
    Game.Tick();
    _resizeTickTimer.Enabled = true;
}

private void OnResizeEnd(object sender, EventArgs eventArgs)
{
    _isResizeTickEnabled = false;
    _resizeTickTimer.Enabled = false;
...
```
And the timer is initialised in the constructor as follows:
```csharp
_resizeTickTimer = new System.Timers.Timer(1) { SynchronizingObject = Form, AutoReset = false };
_resizeTickTimer.Elapsed += OnResizeTick;
```

**Comment by Jjagg** [comment]
@MichaelDePiazzi Did your changes work out?

**Comment by MichaelDePiazzi** [comment]
@Jjagg Yes, it has worked quite well for me. This logic has been in the public release of my game for several months now and there doesn't seem to be any issues with it. It has also proven to be very useful for video playback as the timing for that seems to mess up if updates fall behind.

Note that update timing won't be accurate while this is engaged. But the important thing is that updates are still happening and not falling behind.

If you are good with this as a solution, then I'd be happy to submit a PR to get these changes merged in.

**Comment by Jjagg** [comment]
@MichaelDePiazzi Yes, that would be great! :)

**Comment by MichaelDePiazzi** [comment]
@Jjagg No problem, I'll get on to this as soon as I can.

**Comment by MichaelDePiazzi** [comment]
@Jjagg Ok, I've just submitted PR #6594 which works around this issue on the Windows DirectX build.

The DesktopGL build will need a different fix though, so this issue is still not fully resolved. I'm not familiar with the DesktopGL build or SDL though, so I'm not sure how to do a similar workaround for it.

**Comment by MichaelDePiazzi** [comment]
@Jjagg This issue should get the "OpenGL" label added. And the "DirectX" one can probably be removed since that's resolved now.

**Comment by briermay** [comment]
is there still no fix for this ? I'm using windows 10 and MonoGame 3.7 and Cross platform desktop project upon start locks up still. and as for someone saying XNA had the same behavior, no it does not I have XNA on my laptop for building content (i really dislike the way MonoGame does it) and it does not exhibit the same behavior (although its Direct X not OpenGL)

**Comment by domi1819** [comment]
I've had partial success with hacking MonoGame to initialize the GraphicsDevice and run the update and draw loops on a new thread, while keeping input polling and window events on the main thread.
That way the game kept running even while the window was being moved and resized and at a first glance it seemed to work fine.
However when I tried to switch the render target (draw-to-texture) the SpriteBatcher would just stop working and to this day I haven't been able to figure out what went wrong.
From my research it seems totally possible to run all SDL drawing on a non-main thread but I'm lacking the knowledge to make it work for MonoGame.

**Comment by Apostolique** [comment]
I took some time to do a proof of concept for getting this to work in Windows DesktopGL: <https://github.com/Apostolique/MonoGameContinuousUpdate>.

I used <https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-settimer> to resume updates on the UI thread. The timer is used to apply [this solution](https://gamedev.stackexchange.com/questions/51376/how-do-i-prevent-xna-from-pausing-updates-while-the-window-is-being-resized-move).

I also plug into the SDL event filters to update the MonoGame window size.

There's a bit of extra fluff code with the window maximize that can be ignored.

**Comment by FakeDomi** [comment]
I guess I should write a follow-up here too.
For learning purposes, I'm making my own engine and I tried moving the game / drawing loop off-thread there too.
It works well, the timing is completely unaffected by the user moving / resizing the window.
Unfortunately, I think this approach isn't fit for Monogame since it requires quite some architectural changes to deal with off-thread game logic, and more importantly it could break existing games that assume everything is happening on the UI thread.
